File: src/types/wordbook.ts
import { Id, Timestamp } from './common';

/// 主题标签
export interface ThemeTag {
  id: Id;
  name: string;
  icon: string;
  color: string;
  created_at: Timestamp;
}

/// 单词本
export interface WordBook {
  id: Id;
  title: string;
  description: string;
  icon: string;
  icon_color: string;
  total_words: number;
  linked_plans: number;
  created_at: Timestamp;
  updated_at: Timestamp;
  last_used: Timestamp;
  deleted_at?: Timestamp;
  status: string;
  theme_tags?: ThemeTag[];
}

/// 创建单词本请求
export interface CreateWordBookRequest {
  title: string;
  description: string;
  icon: string;
  icon_color: string;
  theme_tag_ids?: Id[];
}

/// 更新单词本请求
export interface UpdateWordBookRequest {
  title?: string;
  description?: string;
  icon?: string;
  icon_color?: string;
  status?: string;
  theme_tag_ids?: Id[];
}

/// 单词本查询参数
export interface WordBookQuery {
  keyword?: string;
  icon_color?: string;
}

/// 单词
export interface Word {
  id: Id;
  word: string;
  meaning: string;
  description?: string;
  ipa?: string;
  syllables?: string;
  phonics_segments?: string;
  image_path?: string;
  audio_path?: string;
  part_of_speech?: string;
  category_id?: Id;
  word_book_id?: Id;
  // 新增自然拼读分析字段
  pos_abbreviation?: string;
  pos_english?: string;
  pos_chinese?: string;
  phonics_rule?: string;
  analysis_explanation?: string;
  created_at: Timestamp;
  updated_at: Timestamp;
}

/// 创建单词请求
export interface CreateWordRequest {
  word: string;
  meaning: string;
  description?: string;
  ipa?: string;
  syllables?: string;
  phonics_segments?: string;
  part_of_speech?: string;
  category_id?: Id;
  // 新增自然拼读分析字段
  pos_abbreviation?: string;
  pos_english?: string;
  pos_chinese?: string;
  phonics_rule?: string;
  analysis_explanation?: string;
}

/// 更新单词请求
export interface UpdateWordRequest {
  word?: string;
  meaning?: string;
  description?: string;
  ipa?: string;
  syllables?: string;
  phonics_segments?: string;
  part_of_speech?: string;
  category_id?: Id;
  // 新增自然拼读分析字段
  pos_abbreviation?: string;
  pos_english?: string;
  pos_chinese?: string;
  phonics_rule?: string;
  analysis_explanation?: string;
}

/// 单词查询参数
export interface WordQuery {
  keyword?: string;
  difficulty_level?: number;
  category_id?: Id;
  part_of_speech?: string;
}

/// 单词分类
export interface Category {
  id: Id;
  name: string;
  description?: string;
  color: string;
  icon: string;
  word_count: number;
  created_at: Timestamp;
}

/// 单词本统计
export interface WordBookStatistics {
  total_books: number;
  total_words: number;
  word_types: WordTypeDistribution;
}

/// 单词类型分布
export interface WordTypeDistribution {
  nouns: number;
  verbs: number;
  adjectives: number;
  others: number;
}

/// 批量导入单词请求
export interface ImportWordsRequest {
  book_id: Id;
  words: CreateWordRequest[];
  overwrite_existing: boolean;
}

/// AI分析的单词信息
export interface AnalyzedWord {
  word: string;
  meaning: string;
  phonetic?: string;
  part_of_speech?: string;
  example_sentence?: string;
  // 新增自然拼读分析字段
  ipa?: string;
  syllables?: string;
  pos_abbreviation?: string;
  pos_english?: string;
  pos_chinese?: string;
  phonics_rule?: string;
  analysis_explanation?: string;
}

/// 文本分析结果
export interface TextAnalysisResult {
  words: AnalyzedWord[];
  total_count: number;
  difficulty_distribution: Record<number, number>;
  suggested_title: string;
  suggested_description: string;
}

/// 从分析结果创建单词本的请求
export interface CreateWordBookFromAnalysisRequest {
  title: string;
  description: string;
  icon?: string;
  icon_color?: string;
  words: AnalyzedWord[];
  status?: string;
  book_id?: Id; // 如果提供，则向现有单词本添加单词；否则创建新单词本
}

/// 单词保存结果统计
export interface WordSaveResult {
  book_id: Id;
  added_count: number;
  updated_count: number;
  skipped_count: number;
}

/// 单词提取模式
export type WordExtractionMode = 'all' | 'focus';

/// 分析进度状态
export interface AnalysisProgress {
  status: string;           // "analyzing", "completed", "error"
  current_step: string;     // 当前步骤描述
  chunks_received: number;  // 已接收的chunk数量
  total_chars: number;      // 已接收的总字符数
  elapsed_seconds: number;  // 已用时间（秒）
  error_message?: string;   // 错误信息
}

/// 单词本统计
export interface WordBookStatistics {
  total_books: number;
  total_words: number;
  word_types: WordTypeDistribution;
}

/// 单词类型分布
export interface WordTypeDistribution {
  nouns: number;
  verbs: number;
  adjectives: number;
  others: number;
}

File: src-tauri/src/types/wordbook.rs
use serde::{Deserialize, Serialize};
use super::{Id, Timestamp};

/// 主题标签
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ThemeTag {
    pub id: Id,
    pub name: String,
    pub icon: String,
    pub color: String,
    pub created_at: Timestamp,
}

/// 单词本
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct WordBook {
    pub id: Id,
    pub title: String,
    pub description: String,
    pub icon: String,
    pub icon_color: String,
    pub total_words: i32,
    pub linked_plans: i32,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
    pub last_used: Timestamp,
    pub deleted_at: Option<Timestamp>,
    pub status: String,
    pub theme_tags: Option<Vec<ThemeTag>>,
}

/// 创建单词本请求
#[derive(Debug, Serialize, Deserialize)]
pub struct CreateWordBookRequest {
    pub title: String,
    pub description: String,
    pub icon: String,
    pub icon_color: String,
    pub theme_tag_ids: Option<Vec<Id>>,
}

/// 更新单词本请求
#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateWordBookRequest {
    pub title: Option<String>,
    pub description: Option<String>,
    pub icon: Option<String>,
    pub icon_color: Option<String>,
    pub status: Option<String>,
    pub theme_tag_ids: Option<Vec<Id>>,
}

/// 单词本查询参数
#[derive(Debug, Serialize, Deserialize)]
pub struct WordBookQuery {
    pub keyword: Option<String>,
    pub icon_color: Option<String>,
}

/// 单词
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Word {
    pub id: Id,
    pub word: String,
    pub meaning: String,
    pub description: Option<String>,
    pub ipa: Option<String>,
    pub syllables: Option<String>,
    pub phonics_segments: Option<String>,
    pub image_path: Option<String>,
    pub audio_path: Option<String>,
    pub part_of_speech: Option<String>,
    pub category_id: Option<Id>,
    pub word_book_id: Option<Id>,
    // 新增自然拼读分析字段
    pub pos_abbreviation: Option<String>,
    pub pos_english: Option<String>,
    pub pos_chinese: Option<String>,
    pub phonics_rule: Option<String>,
    pub analysis_explanation: Option<String>,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}

/// 创建单词请求
#[derive(Debug, Serialize, Deserialize)]
pub struct CreateWordRequest {
    pub word: String,
    pub meaning: String,
    pub description: Option<String>,
    pub ipa: Option<String>,
    pub syllables: Option<String>,
    pub phonics_segments: Option<String>,
    pub part_of_speech: Option<String>,
    pub category_id: Option<Id>,
    // 新增自然拼读分析字段
    pub pos_abbreviation: Option<String>,
    pub pos_english: Option<String>,
    pub pos_chinese: Option<String>,
    pub phonics_rule: Option<String>,
    pub analysis_explanation: Option<String>,
}

/// 更新单词请求
#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateWordRequest {
    pub word: Option<String>,
    pub meaning: Option<String>,
    pub description: Option<String>,
    pub ipa: Option<String>,
    pub syllables: Option<String>,
    pub phonics_segments: Option<String>,
    pub part_of_speech: Option<String>,
    pub category_id: Option<Id>,
    // 新增自然拼读分析字段
    pub pos_abbreviation: Option<String>,
    pub pos_english: Option<String>,
    pub pos_chinese: Option<String>,
    pub phonics_rule: Option<String>,
    pub analysis_explanation: Option<String>,
}

/// 单词查询参数
#[derive(Debug, Serialize, Deserialize)]
pub struct WordQuery {
    pub keyword: Option<String>,
    pub difficulty_level: Option<i32>,
    pub category_id: Option<Id>,
    pub part_of_speech: Option<String>,
}

/// 单词分类
#[derive(Debug, Serialize, Deserialize)]
pub struct Category {
    pub id: Id,
    pub name: String,
    pub description: Option<String>,
    pub color: String,
    pub icon: String,
    pub word_count: i32,
    pub created_at: Timestamp,
}

/// 单词本统计
#[derive(Debug, Serialize, Deserialize)]
pub struct WordBookStatistics {
    pub total_books: i32,
    pub total_words: i32,
    pub word_types: WordTypeDistribution,
}

/// 单词类型分布
#[derive(Debug, Serialize, Deserialize)]
pub struct WordTypeDistribution {
    pub nouns: i32,
    pub verbs: i32,
    pub adjectives: i32,
    pub others: i32,
}

/// 批量导入单词请求
#[derive(Debug, Serialize, Deserialize)]
pub struct ImportWordsRequest {
    pub book_id: Id,
    pub words: Vec<CreateWordRequest>,
    pub overwrite_existing: bool,
}

/// AI分析的单词信息
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct AnalyzedWord {
    pub word: String,
    pub meaning: String,
    pub part_of_speech: Option<String>,
    pub example_sentence: Option<String>,
    // 新增自然拼读分析字段
    pub ipa: Option<String>,
    pub syllables: Option<String>,
    pub pos_abbreviation: Option<String>,
    pub pos_english: Option<String>,
    pub pos_chinese: Option<String>,
    pub phonics_rule: Option<String>,
    pub analysis_explanation: Option<String>,
    pub word_frequency: Option<i32>,
}

/// 从分析结果创建单词本的请求
#[derive(Debug, Serialize, Deserialize)]
pub struct CreateWordBookFromAnalysisRequest {
    pub title: String,
    pub description: String,
    pub icon: Option<String>,
    pub icon_color: Option<String>,
    pub words: Vec<AnalyzedWord>,
    pub status: Option<String>,
    pub book_id: Option<Id>, // 如果提供，则向现有单词本添加单词；否则创建新单词本
    pub theme_tag_ids: Option<Vec<Id>>, // 主题标签ID列表
}
